Libraries/LibGUI/GTextEditor.cpp

```diff
@@ -111,6 +118,82 @@ void GTextEditor::update_content_size()
     set_size_occupied_by_fixed_elements({ ruler_width(), 0 });
 }
 
+void GTextEditor::update_line_wrapping()
```

Comment:
Forgive me if I misunderstand, because it's early and I haven't had my coffee yet :)

This wrapping looks destructive: it takes m_lines, and alters it. That doesn't seem like what you really want -- typically you just want to change the presentation of the text, not how it is fundamentally structured.

If I load a file, and save it again, I wouldn't expect it to have more lines than when I loaded it. I'd just expect to see single long lines pulled around on screen, with the same line _number_, like this:

![image](https://user-images.githubusercontent.com/125863/61576906-4cd5ff80-aae0-11e9-83c1-dc37b6003465.png)

So I think how this needs to work is to instead wrap on _render_, starting at first_visible_line, and breaking long lines into multiple (rendered!) lines without affecting the numbering or m_lines.

Reply:
Yes, this needs to work very differently.

Each `TextEditor::Line` can produce an arbitrary number of visual lines. The visual lines need to be recomputed whenever the `GTextEditor` changes size, and also when the text content is altered.

`GTextEditor::paint_event()` should operate on these visual lines, rather than on the "real" lines which is what we would write to disk when saving. All of the scrolling, selection and cursor logic also needs to operate on visual lines.

Reply:
Thanks for the advice guys! I'm glad I asked before I dug myself even deeper :)